### PR TITLE
bugfix: make namespace directive not churn unique types

### LINKED ISSUE #5489: Namespace directive breaks idempotency check for unique types
``` unison
namespace foo

type Foo = Foo Nat
toNat  = cases Foo n -> n
```

```
@systemfw/bug/main> 

  Loading changes detected in ~/projects/unison/scratch.u.


  I found and typechecked these definitions in ~/projects/unison/scratch.u. If you do an `add` or
  `update`, here's how your codebase would change:
  
    ⍟ These new definitions are ok to `add`:
    
      type foo.Foo
      foo.toNat : Foo -> Nat

@systemfw/bug/main> add

  ⍟ I've added these definitions:
  
    type foo.Foo
    foo.toNat : Foo -> Nat
```

```
@systemfw/bug/main> load scratch.u

  Loading changes detected in scratch.u.

  I found and typechecked these definitions in scratch.u. If you do an `add` or `update`, here's
  how your codebase would change:
  
    ⍟ These names already exist. You can `update` them to your new definition:
    
      type foo.Foo
      foo.toNat : Foo -> Nat
```

### PR DESCRIPTION
**Pull request by mitchellwrosen**
## Overview

Fixes #5489 

This PR tweaks decl parsing to delay assigning a unique type guid until after the namespace pragma is applied. Previously (as seen in first transcript commit), one could accidentally be assigned a new unique type guid in the presence of a namespace directive.

## Test coverage

I've added a transcript that demonstrates the fix.

### CODE REVIEW

**Comment by aryairani** [comment]
@mitchellwrosen 
> 💥  annotations.refs.Test annotations for types with effects.typechecked file

File: unison-src/transcripts/idempotent/duplicate-names.md [hunk 1]
```diff
@@ -58,7 +58,7 @@ structural ability X where
 ``` ucm :added-by-ucm
   Loading changes detected in scratch.u.
 
-  I found two types called X:
```
Comment (mitchellwrosen) [inline unison-src/transcripts/idempotent/duplicate-names.md:61]: This change is due to the fact that our parser has two different error constructors that represent the duplicate type names. This PR performed only minimal unification (switching one over to the other), but there's a little bit more cleanup to do, namely delete the other one entirely.

**Comment by mitchellwrosen** [comment]
@aryairani Out of expedience I decided to leave the `UniqueName` generation alone, since I realized we could reuse the namespace field in the parser state that was already there for a different purpose, but I can make a separate ticket for refactoring it to be the `gen -> (Bytes, gen)` that we talked about if you'd like.

**Comment by aryairani** [comment]
> @aryairani Out of expedience I decided to leave the `UniqueName` generation alone, since I realized we could reuse the namespace field in the parser state that was already there for a different purpose, but I can make a separate ticket for refactoring it to be the `gen -> (Bytes, gen)` that we talked about if you'd like.

I read over the code but I didn't get the new strategy. If you're convinced it will work, that's fine; if you want to discuss it let me know.

**Review by aryairani — APPROVED** [review]

**Comment by mitchellwrosen** [comment]
@aryairani Yeah I'm confident it will work.

[When we query the parser environment for whether a unique type already exists for a particular name](https://github.com/unisonweb/unison/pull/5509/files#diff-f052a05928de0c067b0ed69b2ca8b529ec915bc4bf66e492421104c4c8165dabR191), we prefix the given name with the current namespace directive, if any.

It's really that simple, I didn't think of this solution right away though. :(

**Pull request merged**